core/validation/utils.py

```diff
@@ -12,3 +15,50 @@ def remove_duplicate_indexes(df: pd.DataFrame):
     :return: the DataFrame without duplicate indexes
     """
     return df[~df.index.duplicated(keep="first")]
+
+
+def numerical_eval(value: Any) -> int | float | Any:
+    """Evaluate value as an int or float. If impossible, return the original value.
+
+    :param value: any value
+    :return: the value as an int or a float
+    :raises ValueError: if value cannot be evaluated as an int or a float
+    """
+    try:
+        return int(value)
+    except ValueError:
+        return float(value)
+
+
+def is_numerical(value: Any) -> bool:
+    """Returns True if the value is numerical, otherwise False.
+
+    :param value: value to check is numerical
+    :return: True if numerical, else False
+    """
+    try:
+        numerical_eval(value)
+        return True
+    except ValueError:
+        return False
+
```

Comment:
Can't we use ```isinstance(value, Number)``` here?

Reply:
The value returned from the df is a string regardless of the value, because the `Response` column includes both text and numerical answers. For that reason this function just tries to to evaluate the `value` to a `float` or `int`, if it can't then it must not be a number and is rejected

Reply:
In that case I think you can just do ```return value.isdecimal()``` as that will return True if the string can be a float or int, and removes the need for the try/catch block.

Reply:
Ah no, forget what I just wrote. That won't work on a float.

Reply:
hmmm that almost works.. Could use this instead though and just add a type hint that the function only works with strings. I think I prefer this


```
if string.replace(".", "").isnumeric():
    return True
else:
    return False
```

Reply:
Won't work for negative numbers though. If we don't expect any of them then I agree this is an improve over the try/catch!

Reply:
Ah good point. Maybe we'll just go with the casting to types option then.. I'll revisit monday

Reply:
Well, that or

```return string.replace(".", "").strip("-").isnumeric()```

Saves using two functions, and two try/catch blocks...but will validate "1000.000.000", which I suppose is a bigger concern? 

```return string.replace(".", "", 1).strip("-").isnumeric()```

Should work in all cases

Reply:
Although that will validate both "10000-" and "10000."

Maybe better to stick with the try/catch blocks, just surprised there's not a more elegant way of evaluating if a string can be a float or an int Python

Reply:
Yep going to stick with the try catch. I agree - it's surprising but I can't see anything online that looks better